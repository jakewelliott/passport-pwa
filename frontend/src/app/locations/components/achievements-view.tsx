import type { Park, ParkActivity } from '@/lib/mock/types';
import { FaRegCheckSquare, FaRegSquare, FaStamp } from 'react-icons/fa';

interface AchievementsViewProps {
  parkActivity: ParkActivity;
  park: Park;
}

const AchievementsView = ({ parkActivity, park }: AchievementsViewProps) => {
  return (
    <div className='flex flex-col gap-3' data-testid='achievements-view'>
      {parkActivity.stampCollectedAt !== undefined && (
        <div className='top-0 flex'>
          <FaStamp size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
          <p>
            {parkActivity.stampCollectedAt
              ? `Stamp collected ${parkActivity.stampCollectedAt}`
              : 'Stamp not yet collected'}
          </p>
        </div>
      )}
      {
        park.bucketListItems &&
        park.bucketListItems.length > 0 && (
          <div className='top-0 flex'>
            {parkActivity.completedBucketListItems[0]?.id ? (
              <FaRegCheckSquare size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
            ) : (
              <FaRegSquare size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
            )}
            <p data-testid={'BLI'}>
              Bucket List Item:
              <br />
              {park.bucketListItems[0].task}
            </p>
          </div>
        )}
    </div>
  );
};

export default AchievementsView;
