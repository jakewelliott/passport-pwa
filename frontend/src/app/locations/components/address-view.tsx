import type { Address, Park } from '@/types';
import { useState } from 'react';
import { FiNavigation } from 'react-icons/fi';


const ParkAddress = ({ address }: { address: Address }) => {
	return (
		<div>
			<p>{address.title}</p>
			<p>{address.addressLineOne}</p>
			{address.addressLineTwo && <p>{address.addressLineTwo}</p>}
			<p>{address.city}, {address.state} {address.zipcode}</p>
		</div>
	);
};

export const AddressView = ({ park }: { park: Park }) => {

	const [showAll, setShowAll] = useState(false);
	const showButton = park.addresses && park.addresses.length > 1;

	const toggle = () => {
		setShowAll(!showAll);
	};

	const OtherAddresses = () =>
		park.addresses.slice(1).map((address) => (
			<ParkAddress key={address.title} address={address} />
		))


	const Button = () =>
		<button type='button' onClick={toggle} className='ml-4 inline-flex items-center'>
			{showAll ? 'Show Less' : 'Show More'}
		</button>

	return (
		<div className='top-0 flex' key={park.parkName}>
			<FiNavigation size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<ParkAddress address={park.addresses[0]} />
			{showAll && <OtherAddresses />}
			{showButton && <Button />}
		</div>
	);
};
