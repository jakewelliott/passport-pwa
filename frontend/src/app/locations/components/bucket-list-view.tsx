import { useBucketList } from '@/hooks/queries/useBucketList';
import { useStamp } from '@/hooks/queries/useStamps';
import DateHelper from '@/lib/date-helper';
import type { BucketListItem, CollectedStamp, Park } from '@/lib/mock/types';
import { FaRegCheckSquare, FaRegSquare, FaStamp } from 'react-icons/fa';

const BucketListItemView = ({ bucketListItem }: { bucketListItem: BucketListItem }) => {
	const isCompleted = bucketListItem.completedAt && !bucketListItem.deleted;
	const Checkbox = isCompleted ? FaRegCheckSquare : FaRegSquare;

	return (
		<div>
			<Checkbox size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<p data-testid={'BLI'}>
				Bucket List Item:
				<br />
				{bucketListItem.task}
			</p>
		</div>
	);
};

const StampView = ({ stamp }: { stamp: CollectedStamp | undefined }) => {
	return (
		<>
			<FaStamp size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<p>
				{stamp
					? `Stamp collected on ${DateHelper.stringify(new Date(stamp.createdAt)).replace(',', ' at')}`
					: 'Stamp not yet collected'}
			</p>
		</>
	);
};

export const BucketListView = ({ park }: { park: Park }) => {
	const { data: items, isLoading: itemsLoading } = useBucketList(park.id);
	const { data: stamp, isLoading: stampLoading } = useStamp(park.abbreviation);

	// TODO: replace with a placeholder
	if (itemsLoading || stampLoading) return null;

	return (
		<div className='flex flex-col gap-3' data-testid='achievements-view'>
			<div className='top-0 flex'>
				<StampView stamp={stamp} />
			</div>
			<div className='top-0 flex'>
				{items.map((item: BucketListItem) => (
					<BucketListItemView key={item.id} bucketListItem={item} />
				))}
			</div>
		</div>
	);
};
