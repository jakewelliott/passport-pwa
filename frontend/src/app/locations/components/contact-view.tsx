import type { Park } from '@/types';
import { FaRegEnvelope } from 'react-icons/fa';
import { FiNavigation, FiPhone } from 'react-icons/fi';

export const ContactView = ({ park }: { park: Park }) => {
    return (
        <div className='mx-4 flex flex-col gap-3' data-testid='location-contact'>
            {/* Coordinates */}
            <div className='top-0 flex'>
                <FiNavigation size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>
                    GPS: {park.coordinates.latitude.toFixed(4)}, {park.coordinates.longitude.toFixed(4)}
                </p>
            </div>
            {/* Phone */}
            <div className='top-0 flex'>
                <FiPhone size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>{String(park.phone).replace(/(\d{3})(\d{3})(\d{4})/, '($1) $2-$3')}</p>
            </div>
            {/* Email */}
            <div className='top-0 flex'>
                <FaRegEnvelope size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>{park.email}</p>
            </div>
        </div>
    );
};
