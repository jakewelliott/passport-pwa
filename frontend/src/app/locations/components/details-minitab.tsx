import type { Park } from '@/types';
import React from 'react';

export const DetailsMiniTab = ({ park }: { park: Park }) => {
	const renderTrails = (trails: string) => {
		return trails
			.split('\n')
			.filter((line) => line.trim() !== '') // Remove empty lines
			.map((line, index, array) => (
				<React.Fragment key={line}>
					{line}
					{index < array.length - 1 && <br />} {/* Add <br /> only if it's not the last line */}
				</React.Fragment>
			));
	};

	return (
		<div className='mt-6 mb-6 flex flex-col'>
			<p className='pr-6 pl-6'>
				{park.establishedYear && (
					<>
						<span className='green-text'>Established: </span>
						{park.establishedYear}
						<br />
						<br />
					</>
				)}
				{park.landmark && (
					<>
						<span className='green-text'>Landmark: </span>
						{park.landmark}
						<br />
						<br />
					</>
				)}
				{park.youCanFind && (
					<>
						<span className='green-text'>You can find... </span>
						{park.youCanFind}
						<br />
						<br />
					</>
				)}
				{park.trails && (
					<>
						<span className='green-text'>Trails: </span>
						<br />
						{renderTrails(park.trails)}
						<br />
						<br />
					</>
				)}
			</p>
			<div data-testid='icon-scroll-container' className='icon-scroll-container overflow-x-auto'>
				<div className='inline-flex gap-6 px-6'>
					{park.icons?.map((icon, index) => (
						<img
							src={`/icons/park/${icon.iconName}.svg`}
							width={55}
							height={55}
							key={icon.iconName}
							alt={`Park icon ${index + 1}`}
						/>
					))}
					<div className='w-px flex-shrink-0' />
				</div>
			</div>
		</div>
	);
};
