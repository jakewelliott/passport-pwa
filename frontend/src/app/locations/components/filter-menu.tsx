import { FaFilter } from 'react-icons/fa';

interface FilterMenuProps {
    searchQuery: string;
    setSearchQuery: (value: string) => void;
    setIsFilterModalOpen: (value: boolean) => void;
}

export const FilterMenu = ({ searchQuery, setSearchQuery, setIsFilterModalOpen }: FilterMenuProps) => (
    <div className='relative flex gap-2'>
        <div className='relative flex-1'>
            <input
                type='text'
                placeholder='Search parks...'
                value={searchQuery}
                onChange={(e) => setSearchQuery(e.target.value)}
                className='w-full rounded-lg border border-system-gray bg-supporting-lightblue p-3 pr-7 focus:border-secondary-darkteal focus:outline-none focus:ring-1 focus:ring-secondary-darkteal focus:ring-opacity-100'
            />
            {searchQuery && (
                <button
                    type='button'
                    onClick={() => setSearchQuery('')}
                    className='-translate-y-1/2 absolute top-1/2 right-3 transform bg-supporting-lightblue text-system-gray hover:text-secondary-darkteal'
                    aria-label='Clear search'
                >
                    &times;
                </button>
            )}
        </div>
        <button
            type='button'
            onClick={() => setIsFilterModalOpen(true)}
            className={
                'w-auto rounded-lg border border-system-gray bg-system-white px-3 py-2 transition-all duration-200'
            }
        >
            <FaFilter />
        </button>
    </div>
);
