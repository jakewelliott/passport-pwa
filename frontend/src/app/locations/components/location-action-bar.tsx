import type { Park } from '@/types';
import { FiGlobe, FiMail, FiNavigation, FiPhone } from 'react-icons/fi';

export const LocationActionBar = ({ park }: { park: Park }) => {
    // TODO: styling here needs to be fixed
    return (
        <div
            className='flex flex-row items-center justify-evenly bg-secondary-darkteal py-2.5'
            data-testid='location-action-bar'
        >
            {park.coordinates?.latitude && park.coordinates.longitude && (
                <a
                    href={`https://www.google.com/maps/place/${park.coordinates.latitude},${park.coordinates.longitude}`}
                    className='text-supporting-inactiveblue'
                >
                    <FiNavigation size={24} strokeWidth={3} />
                </a>
            )}
            {park.phone && (
                <a href={`tel://${park.phone}`} className='text-supporting-inactiveblue'>
                    <FiPhone size={24} strokeWidth={3} />
                </a>
            )}
            {park.website && (
                <a href={park.website} className='text-supporting-inactiveblue'>
                    <FiGlobe size={24} strokeWidth={3} />
                </a>
            )}
            {park.email && (
                <a href={`mailto:${park.email}`} className='text-supporting-inactiveblue'>
                    <FiMail size={24} strokeWidth={3} />
                </a>
            )}
        </div>
    );
};
