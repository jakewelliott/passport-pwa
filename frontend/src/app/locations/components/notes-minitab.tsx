import RoundedButton from '@/components/rounded-button';
import { useNote, useUpdateNote } from '@/hooks/queries/useNotes';
import { dbg, dbgif } from '@/lib/debug';
// components/NotesMiniTab.tsx
import { useEffect, useState } from 'react';
import { useBeforeUnload } from 'react-router';
import { toast } from 'react-toastify';

export const NotesMiniTab = ({
    parkId,
}: {
    parkId: number;
}) => {
    const { data: remoteNote, refetch, isLoading } = useNote(parkId);
    const { mutate } = useUpdateNote();

    const [hasUnsavedChanges, setHasUnsavedChanges] = useState(false);
    const [noteState, setNoteState] = useState('');

    useBeforeUnload(() => {
        if (hasUnsavedChanges) {
            return 'You have unsaved changes. Are you sure you want to leave?';
        }
    });

    useEffect(() => {
        dbg('EFFECT', 'NotesMiniTab');
        dbgif(isLoading, 'EFFECT', 'NotesMiniTab', 'Note is still loading!');

        // check and see if we got the note from the server
        if (remoteNote?.note === undefined) {
            dbg('EFFECT', 'NotesMiniTab', 'Note is still loading!');
        } else if (remoteNote.note !== noteState) {
            dbg('EFFECT', 'NotesMiniTab', 'Server note changed, updating state...');
            dbg('EFFECT', 'NotesMiniTab', `Remote note: ${remoteNote.note}`);
            dbg('EFFECT', 'NotesMiniTab', `Note state: ${noteState}`);
            setNoteState(remoteNote.note);
        }
    }, [remoteNote, isLoading]);

    useEffect(() => {
        const handleBeforeUnload = (e: BeforeUnloadEvent) => {
            if (hasUnsavedChanges) {
                e.preventDefault();
                e.returnValue = '';
            }
        };

        window.addEventListener('beforeunload', handleBeforeUnload);

        return () => {
            window.removeEventListener('beforeunload', handleBeforeUnload);
        };
    }, [hasUnsavedChanges]);

    // const blocker = useBlocker(
    //     ({ currentLocation, nextLocation }) => hasUnsavedChanges && currentLocation.pathname !== nextLocation.pathname,
    // );

    // useEffect(() => {
    //     if (blocker.state === 'blocked') {
    //         const proceed = window.confirm('You have unsaved changes. Are you sure you want to leave?');
    //         if (proceed) {
    //             blocker.proceed();
    //         } else {
    //             blocker.reset();
    //         }
    //     }
    // }, [blocker]);

    const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        setNoteState(e.target.value);
        setHasUnsavedChanges(true);
    };

    const handleClick = () => {
        mutate(
            { parkId, note: noteState },
            {
                onSuccess: () => {
                    refetch();
                    toast.success('Notes saved!');
                    setHasUnsavedChanges(false);
                },
                onError: () => {
                    toast.error('Failed to save notes');
                    dbg('ERROR', 'NotesMiniTab', 'Failed to save notes');
                },
            },
        );
    };

    if (isLoading) return <div>Loading...</div>;
    const placeholder = parkId === 0 ? 'Add some general notes!' : 'Add some personal notes about this park!';

    return (
        <div className='flex flex-col items-center gap-4 px-4 pt-4'>
            <textarea
                className='h-72 w-full flex-grow resize-none border border-secondary-darkteal bg-supporting-lightblue p-4 focus:border-secondary-darkteal focus:outline-none focus:ring-1 focus:ring-secondary-darkteal focus:ring-opacity-100'
                value={noteState}
                onChange={handleChange}
                placeholder={placeholder}
            />
            <RoundedButton title={'Save'} onClick={handleClick} />
        </div>
    );
};
