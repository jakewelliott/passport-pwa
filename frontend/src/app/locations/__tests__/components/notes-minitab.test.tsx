import { NotesMiniTab } from '@/app/locations/components/notes-minitab';
import { mockPark, mockParkNote } from '@/lib/testing/mock';
import { renderWithClient } from '@/lib/testing/test-wrapper';
import { fireEvent, screen } from '@testing-library/react';
import { toast } from 'react-toastify';
import { describe, expect, it, vi } from 'vitest';

// TODO: this file is F'ed

// Mock react-toastify
vi.mock('react-toastify', () => ({
    toast: {
        success: vi.fn(),
        error: vi.fn(),
        info: vi.fn(),
        warning: vi.fn(),
    },
}));

describe('NotesMiniTab', () => {
    it('renders the notes textarea with initial value', () => {
        renderWithClient(<NotesMiniTab parkId={mockPark.id} />);
        const textarea = screen.getAllByPlaceholderText('Add some personal notes about this park!');
        expect(textarea).toHaveValue(mockParkNote.note);
    });

    // it('updates note on textarea change', async () => {
    // 	renderWithClient(<NotesMiniTab parkId={mockPark.id} />);
    // 	const textarea = screen.getByRole('textbox');
    // 	const newNote = 'Updated test note';

    // 	fireEvent.change(textarea, { target: { value: newNote } });

    // });

    it('shows success toast when note is saved', async () => {
        renderWithClient(<NotesMiniTab parkId={mockPark.id} />);
        const saveButton = screen.getByTestId('save-button');

        fireEvent.click(saveButton);

        // TODO: make this check for the correct mutation
        // await waitFor(() => {
        // 	expect(mockMutate).toHaveBeenCalledWith(
        // 		{ parkId: mockPark.id, note: mockParkNote.note },
        // 		expect.objectContaining({
        // 			onSuccess: expect.any(Function),
        // 			onError: expect.any(Function),
        // 		}),
        // 	);
        // });

        // Simulate successful save
        // const successCallback = mockMutate.mock.calls[0][1].onSuccess;
        // successCallback();

        expect(toast.success).toHaveBeenCalledWith('Notes saved!');
    });

    it('shows error toast when note save fails', async () => {
        renderWithClient(<NotesMiniTab parkId={mockPark.id} />);
        const saveButton = screen.getByTestId('save-button');

        fireEvent.click(saveButton);

        // TODO: make this check for the correct mutation
        // await waitFor(() => {
        // 	expect(mockMutate).toHaveBeenCalled();
        // });

        // Simulate failed save
        // const errorCallback = mockMutate.mock.calls[0][1].onError;
        // errorCallback();

        expect(toast.error).toHaveBeenCalledWith('Failed to save notes');
    });

    it('displays loading state when data is being fetched', () => {
        renderWithClient(<NotesMiniTab parkId={mockPark.id} />);
        expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('synchronizes local note with remote note when remote note differs', () => {
        const remoteNote = 'Remote note';

        renderWithClient(<NotesMiniTab parkId={mockPark.id} />);

        // expect(mockSetNote).toHaveBeenCalledWith(mockAbbreviation, remoteNote);
    });

    it('does not synchronize when remote note matches local note', () => {
        renderWithClient(<NotesMiniTab parkId={mockPark.id} />);

        // expect(mockSetNote).not.toHaveBeenCalled();
    });

    it('handles null remote note gracefully', () => {
        renderWithClient(<NotesMiniTab parkId={mockPark.id} />);
        // expect(mockSetNote).not.toHaveBeenCalled();
    });

    it('handles empty local note when saving', async () => {
        renderWithClient(<NotesMiniTab parkId={mockPark.id} />);
        const saveButton = screen.getByTestId('save-button');

        fireEvent.click(saveButton);

        // await waitFor(() => {
        // 	expect(mockMutate).toHaveBeenCalledWith({ parkId: mockPark.id, note: '' }, expect.any(Object));
        // });
    });
});
