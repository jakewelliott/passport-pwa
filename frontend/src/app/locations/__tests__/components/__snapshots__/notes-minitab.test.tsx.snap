// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotesMiniTab > matches snapshot 1`] = `
<div>
  <div
    class="flex h-full flex-col"
  >
    <textarea
      class="h-72 w-full flex-grow resize-none border border-secondary_darkteal p-4 focus:border-secondary_darkteal focus:outline-none focus:ring-1 focus:ring-secondary_darkteal focus:ring-opacity-100"
      placeholder="Add some personal notes about this park!"
    />
    <div
      class="flex justify-center p-3"
      data-testid="save-button"
    >
      <button
        class="bg-secondary_darkteal flex cursor-pointer select-none items-center justify-center text-center p-4 rounded-full"
        style="min-width: 128px;"
        type="button"
      >
        <p
          class="text-system_white button-text"
        >
          Save
        </p>
      </button>
    </div>
  </div>
</div>
`;
